File: src/taeeun/week2/1-3.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    stack<int> st; //보조컨테이너
    vector<int> container; //실제 컨테이너
    int box=1;
    for (int i=0; i<order.size(); i++) {
        while (box<=order[i]) {
            st.push(box);
            box++;
        }
        if (st.top()>order[i]) break;
        if (st.top()==order[i]) st.pop();
        container.push_back(order[i]);
    }
    answer=container.size();
    return answer;
}

File: src/geonhyeong/week2/문제집 1/1-3.py
#
#  1-3.swift
#  Question Link:
#  Primary idea:   스택
#
#  Time Complexity : O()
#  Space Complexity : O()
#  Runtime:  ms
#  Memory Usage:  MB
#

def solution(order):
    answer = 0
    stack = []
    i = 0
    num = 1


    while i < len(order):
        if stack:
            if stack[-1] == order[i]:
                answer += 1
                stack.pop()
            elif stack[-1] < order[i]:
                while num <= order[i]:
                    if num == order[i]:
                        answer += 1
                    else:
                        stack.append(num)
                    num += 1
            else:
                break
        else:
            while num <= order[i]:
                if num == order[i]:
                    answer += 1
                else:
                    stack.append(num)
                num += 1
        i += 1
    
    return answer

# print(solution([4, 3, 1, 2, 5])) # 2
# print(solution([5, 4, 3, 2, 1])) # 5
print(solution([3, 4, 1, 2]))

# 시험 본뒤 정리
def solution(order):
    answer = 0
    stack = []
    num = 1

    for (i, o) in enumerate(order):
        if stack and stack[-1] == o:
            answer += 1
            stack.pop()
        else:
            if not stack or stack[-1] < o:
                while num <= o:
                    if num == o:
                        answer += 1
                    else:
                        stack.append(num)
                    num += 1
            elif stack: # stack[-1] >= o
                break
    
    return answer
